include/drivers/can.h

```diff
@@ -355,7 +363,22 @@ static inline int z_impl_can_send(struct device *dev,
 	const struct can_driver_api *api =
 		(const struct can_driver_api *)dev->driver_api;
 
-	return api->send(dev, msg, timeout, callback_isr, callback_arg);
+	if (callback_isr) {
+		return api->send(dev, msg, timeout, callback_isr, callback_arg);
+	} else {
+		struct can_send_wait send_wait;
+		int ret;
+
+		k_sem_init(&send_wait.sem, 0, 1);
```

Comment:
This change will introduce an overhead of initialising a semaphore on each `can_send()` call without any callback defined. I am not sure it is worth it.

It also seem very unlike any other API implementation in Zephyr?

Reply:
Yea I am sort of ambivalent about it. Not sure how much of a trade-off the memory saved vs the calling overhead it is. 
In terms of driver complexity I think it isn't a big improvement, we all had the same approach and it's not so hard to follow either.

Reply:
>This change will introduce an overhead of initialising a semaphore on each can_send() call without any callback defined. I am not sure it is worth it.

The overhead of initializing is not that much. The blocking send call is used rarely, and in the current implementation, we need to maintain a sem for every mailbox just in case we will need it.

>In terms of driver complexity I think it isn't a big improvement, we all had the same approach and it's not so hard to follow either.

The complexity raised with the timeout patch :-)
I really don't like #19707 in its current state.

Reply:
Ah gotcha. What would be your plan then, use that can_common_isr_callback_handler() for failing sends with timeouts?

Reply:
Updated #19707 on top of this PR. Looks better now!